atomic_reactor/util.py

```diff
@@ -1672,3 +1672,207 @@ def chain_get(d, path, default=None):
+    source and organization with set_source_registry_and_org method
+    and local tags in _local_parents via setter with valid key
+    """
+    def __init__(self, dfp_parents):
+        self._original_parents = deepcopy(dfp_parents) if dfp_parents else []
+        self._custom_parent_image = False
+        self._custom_base_image = False
+        self._base_from_scratch = False
+        self._source_registry = None
+        self._organization = None
+        self._original_base_image = self._original_parents[-1] if self._original_parents else None
+        self._pullable_parents = []
+        self._local_parents = []
+        self._create_pullable()
+        self._source_and_org_set = False
+
+    @property
+    def original_parents(self):
+        """
+        provides unmodified parent_images from Dockerfile
+        :return: list
+        """
+        return self._original_parents
+
+    @property
+    def original_base_image(self):
+        """
+        provides unmodified base image from Dockerfile
+        :return: str or None
+        """
+        return self._original_base_image
+
+    @property
+    def base_image(self):
+        """
+        provides local tag for base_image, or base image with registry and enclosed
+        when we don't know local tag yet
+        :return: str for scratch or ImageName
+        """
+        if self.base_from_scratch:
+            return SCRATCH_FROM
+
+        if self._local_parents and self._local_parents[0]:
+            return deepcopy(self._local_parents[0])
+        else:
+            return self.base_image_key
+
+    @property
+    def base_image_key(self):
+        """
+        provides base image with registry and enclosed
+        (original base image with added registry and enclosed)
+        :return: str for scratch or ImageName
+        """
+        if self.base_from_scratch:
+            return SCRATCH_FROM
+
+        if self._pullable_parents and self._pullable_parents[0]:
+            return deepcopy(self._pullable_parents[0])
+        else:
+            raise KeyError('base_image')
+
+    @property
+    def base_from_scratch(self):
+        """
+        base image is FROM scratch
+        :return: boolean
+        """
+        return self._base_from_scratch
+
+    @property
+    def custom_base_image(self):
+        """
+        base image is custom (FROM koji/image-build)
+        :return: boolean
+        """
+        return self._custom_base_image
+
+    @property
+    def custom_parent_image(self):
+        """
+        any parent image is custom (FROM koji/image-build)
+        :return: boolean
+        """
+        return self._custom_parent_image
+
+    @property
+    def source_registry(self):
+        return self._source_registry
+
+    @property
+    def organization(self):
+        return self._organization
+
+    def __len__(self):
+        return len(self._pullable_parents)
+
+    def __setitem__(self, key, item):
```

Comment:
Could we also have docstrings for `__getitem__` and `__setitem__` to explain that both enclosed and un-enclosed keys will work (and under what conditions, e.g. if the key already has a different registry, it won't work)?

Reply:
sure

Reply:
added